agent/agent.go

```diff
@@ -105,31 +162,91 @@ func (a *agentService) Authenticate(req *http.Request) error {
 	return nil
 }
 
-func (a *agentService) Proxy() *httputil.ReverseProxy {
-	target, err := url.Parse(a.config.OllamaURL)
-	if err != nil {
-		log.Printf("Invalid Ollama URL: %v", err)
+func (a *agentService) Proxy() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		targetURL, err := a.determineTarget(r)
+		if err != nil {
+			log.Printf("Failed to determine target: %v", err)
+			http.Error(w, "Bad Gateway", http.StatusBadGateway)
+
+			return
+		}
+
+		target, err := url.Parse(targetURL)
+		if err != nil {
+			log.Printf("Invalid target URL %s: %v", targetURL, err)
+			http.Error(w, "Bad Gateway", http.StatusBadGateway)
+
+			return
+		}
+
+		proxy := httputil.NewSingleHostReverseProxy(target)
+		proxy.Transport = a.transport
+
+		originalDirector := proxy.Director
+		proxy.Director = func(req *http.Request) {
+			originalDirector(req)
+			a.modifyHeaders(req)
+			log.Printf("Agent forwarding to %s: %s %s", targetURL, req.Method, req.URL.Path)
+		}
+
+		proxy.ErrorHandler = func(w http.ResponseWriter, _ *http.Request, err error) {
+			log.Printf("Proxy error: %v", err)
+			http.Error(w, "Bad Gateway", http.StatusBadGateway)
+		}
 
-		return nil
+		proxy.ServeHTTP(w, r)
+	})
+}
+
+func (a *agentService) AddRoute(rule RouteRule) error {
+	if rule.Name == "" {
+		return errors.New("route name is required")
 	}
 
-	proxy := httputil.NewSingleHostReverseProxy(target)
+	if rule.TargetURL == "" {
+		return errors.New("target URL is required")
+	}
 
-	proxy.Transport = a.transport
+	if _, err := url.Parse(rule.TargetURL); err != nil {
+		return fmt.Errorf("invalid target URL: %w", err)
+	}
 
-	originalDirector := proxy.Director
-	proxy.Director = func(req *http.Request) {
-		originalDirector(req)
-		a.modifyHeaders(req)
-		log.Printf("Agent forwarding to Ollama: %s %s", req.Method, req.URL.Path)
+	err := a.RemoveRoute(rule.Name)
+	if err != nil && !errors.Contains(err, errRouteNotFound) {
+		return err
 	}
 
-	proxy.ErrorHandler = func(w http.ResponseWriter, _ *http.Request, err error) {
-		log.Printf("Proxy error: %v", err)
-		http.Error(w, "Bad Gateway", http.StatusBadGateway)
+	a.routes = append(a.routes, rule)
+
+	for i := 0; i < len(a.routes); i++ {
+		for j := i + 1; j < len(a.routes); j++ {
+			if a.routes[j].Priority > a.routes[i].Priority {
+				a.routes[i], a.routes[j] = a.routes[j], a.routes[i]
+			}
+		}
```

Comment:
Same here, do not manually sort. In fact, do not sort at all, but keep the slice sorted by using binary search (also from the `sort` package).